### ISSUE TITLE
Updated features in two graphs mismatch, but they should be the same theoretically

### ISSUE THREAD

**Issue by xcwanAndy** [dmlc/dgl#2647]
## 🐛 Bug

Hi all. I'm making a test on GraphSAGE for research purpose.

I make two experiments with cora dataset. 

The first: I partition the cora dataset into two parts according to the [given metis partition method](https://github.com/dmlc/dgl/blob/0.5.x/docs/source/guide/distributed-preprocessing.rst) and only uses the first partition, which contains 1512 nodes (1354 inner-nodes and 158 one-hop nodes) and 5370 edges. Then I set the 1512 nodes' features (pre-merge those features) as input and run SAGEConv (single layer) and obtain 1512 output features.

The second: I use the whole cora dataset's 2708 nodes but with exactly the same 5370 edges as the first experiment. Then I set the 2708 nodes' features as input and run SAGEConv (single layer, and the same hyperparameters setting as the first one) and obtain 2708 output features.

As I extract the corresponding 1354 inner-nodes' features from the two experiments and compare between each other, I found them not at the same value.

__Important__: To my knowledge, the partition process uses [in_subgraph](https://github.com/dmlc/dgl/blob/master/python/dgl/subgraph.py#L303) function, which implies that all nodes within a partitioned graph is responsible for all inner-nodes' one-hop update, is that correct? If yes, then I believe the two experiments should output exactly the same value according to the MPNN abstraction (because all necessary nodes and edges are already included in the partitioned subgraph + one-hop nodes' features).

## To Reproduce

1. For the first experiment, just partition the graph using [metis partition](https://github.com/dmlc/dgl/blob/master/examples/pytorch/graphsage/experimental/partition_graph.py) and load the first partition. Then attach the 158 one-hop nodes with their corresponding features. Then set the 1512 features as input and process SAGEConv (you may refer to the parameter setting of this function in `train_sampling.py`), and finally obtain the 1354 inner-nodes' features.

2. For the second experiment, you may run `train_sampling.py` in [folder](https://www.dropbox.com/sh/7q1ym2kys54j6nz/AADt-5YM7dfexJiKnwpGdQJWa?dl=0) and obtain the 1354 inner-nodes' features accordingly.

3. Finally, compare the two output tensors and use `torch.all(torch.eq(a, b))` for comparison.

The result is shown here:
```
Not equal
(tensor([  31,   46,   46,   57,   57,   57,   64,   64,  124,  150,  150,  157,
         157,  157,  157,  202,  202,  202,  202,  205,  222,  222,  222,  228,
         228,  228,  240,  272,  272,  277,  277,  331,  381,  405,  405,  405,
         405,  405,  413,  488,  488,  488,  488,  518,  518,  518,  533,  554,
         554,  554,  669,  681,  742,  742,  742,  758,  762,  762,  762,  762,
         955,  955,  955,  959,  963,  963,  981,  997,  997,  997, 1003, 1009,
        1009, 1073, 1073, 1073, 1073, 1073, 1085, 1085, 1085, 1085, 1085, 1085,
        1090, 1090, 1090, 1090, 1142, 1142, 1142, 1142, 1173, 1173, 1182, 1220]), tensor([ 0,  5,  7,  2,  5,  7,  8, 11,  5,  8,  9,  2,  5, 12, 15,  0,  6,  8,
         9, 13,  9, 10, 12,  5, 11, 12,  9, 10, 11,  4, 11,  4,  8,  0,  6,  8,
         9, 14,  6,  9, 10, 11, 14,  5,  6, 11,  6,  4,  7, 11,  0,  5,  4,  8,
        12,  0,  1,  3,  4,  5,  2,  5,  6, 10,  5, 10, 14,  4,  5, 14, 10,  0,
        10,  0,  1,  6, 10, 12,  1,  4,  5,  6,  8,  9,  0,  5,  8,  9,  4,  5,
        10, 14,  6, 10,  2,  4]))
```
## Expected behavior
The two tensors are equal.

## Environment

 - DGL Version: 0.5.2
 - Backend Library & Version (e.g., PyTorch 0.4.1, MXNet/Gluon 1.3): PyTorch 1.6.0
 - OS (e.g., Linux): Linux
 - How you installed DGL (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source):
 - Python version: 3.7.0
 - CUDA/cuDNN version (if applicable): 10.2.89/7.6.5
 - GPU models and configuration (e.g. V100): V100
 - Any other relevant information: No sampling (set the fan-outs to -1)

**Comment by BarclayII** [comment]
Several questions from my side:
* I noticed that there are two output tensors in your result.  What are they and which statement/function in your code generates them?  They don't seem like output tensors from SAGEConv as they shouldn't be integers otherwise.
* Are the two `train_sampling.py` referring to the same file you shared?
* Are the model parameters the same across experiments?
* Looks like you are using neighbor sampling with fanouts -1.  Are the output nodes the same across experiments for each batch?
* What's the difference between the two output tensors?  If they are small, that might as well be floating point errors.

**Comment by xcwanAndy** [comment]
Thanks for your reply. FYI,
> * I noticed that there are two output tensors in your result.  What are they and which statement/function in your code generates them?  They don't seem like output tensors from SAGEConv as they shouldn't be integers otherwise.

- I stored the output features using `torch.save` function. You may find it in `forward` function
```
            h = layer(block, h)
            # if l != len(self.layers) - 1:
            #     h = self.activation(h)
            #     h = self.dropout(h)

            # TODO(wxc): for debug, store output features
            if self.args.seed != 0:
                f = self.args.log_dir + '/output-feats-' + str(l) + '.pt'
                th.save(h, f)
            print("After h:", h)
```

> * Are the two `train_sampling.py` referring to the same file you shared?

- I have updated both the two experiments' code. You may find the [exp1](https://www.dropbox.com/s/n08lb4nook77hu0/train_sampling_exp1.py?dl=0) and [exp2](https://www.dropbox.com/s/z9dwzem3f7n2z78/train_sampling_exp2.py?dl=0) here. And you may compare their generated output features to verify if they are the same.

> * Are the model parameters the same across experiments?

- Yes, the model parameters are initialized in exactly the same way.

> * Looks like you are using neighbor sampling with fanouts -1.  Are the output nodes the same across experiments for each batch?

- Yes, the output nodes are all the same as I use the same index to compare their labels. And the labels are exactly the same.

> * What's the difference between the two output tensors?  If they are small, that might as well be floating point errors.

- The two experiments' output tensors have different size, as one is 1512 * 1433 and the other is 2708 * 1433. However, I compare the same 1512 nodes' features and find out that they are not in the same value.

**Comment by xcwanAndy** [comment]
@BarclayII

**Comment by BarclayII** [comment]
Could you also share the partition files?

**Comment by xcwanAndy** [comment]
@BarclayII  Yes, you may find the partition files [here](https://www.dropbox.com/sh/a3gjitp7rfb2f48/AADsQKTYvZlBgr7L4L0pxkofa?dl=0).

**Comment by BarclayII** [comment]
> * Yes, the model parameters are initialized in exactly the same way.

"Initialized" the same way may still give you different parameter values since the initializations merely specifies which distribution to draw parameter values from.

I replaced the model parameters with exactly the same values and observed that the results are indeed the same.  You can ensure that the parameters are the same by fixing torch's random seed and reinitializing the parameters with something like
```python
with th.random.fork_rng():
    th.manual_seed(1234)
    for p in model.parameters():
        p.data[:] = th.randn(*p.shape)
```

**Comment by BarclayII** [comment]
Closing this issue.  Please feel free to reopen if things did not work.
